### ISSUE TITLE
Add support for non-array arguments to shard_map

### ISSUE THREAD

**Issue by inailuig** [google/jax#17461]
I think it would be convenient if shard_map allowed for non-array arguments,
right now one has to use a partial to capture those arguments first since the following does not work:

```python
from functools import partial
import jax
import jax.numpy as jnp
from jax.experimental.shard_map import shard_map
from jax.sharding import Mesh, PartitionSpec as P, PositionalSharding


x = jnp.ones(jax.device_count())
x = jax.jit(lambda x: x, out_shardings=PositionalSharding(jax.devices()))(x)
f = lambda x: 2*x+1


mesh = Mesh(jax.devices(), axis_names=("i"))
in_specs = None, P("i")
out_specs = P("i")

@partial(jax.jit, static_argnums=0)
@partial(shard_map, mesh=mesh, in_specs=in_specs, out_specs=out_specs)
def test(f, x):
    return f(x)

test(f, x)
# TypeError: Cannot interpret value of type <class 'function'> as an abstract array; it does not have a dtype attribute
```

**Comment by PhilipVinc** [comment]
@mattjj is this planned at any time or this is very low priority for you?

**Comment by mattjj** [comment]
Sorry for not responding sooner.

#22049 should fix this, and I checked it makes the code in the OP run without error.

**Comment by mattjj** [comment]
Fixed by #22049
